doc/SAI_Proposal_Generic_Extensions.md

```diff
@@ -0,0 +1,423 @@
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_generic_programmable_attribute_fn)(
+        _In_ sai_object_id_t generic_programmable_id,
+        _In_ uint32_t attr_count,
+        _Inout_ sai_attribute_t *attr_list);
+
+/**
+ * @brief Generic extensions methods table retrieved with sai_api_query()
+ **/
+typedef struct _sai_generic_programmable_api_t
+{
+    sai_create_generic_programmable_fn            create_generic_programmable;
+    sai_remove_generic_programmable_fn            remove_generic_programmable;
+    sai_set_generic_programmable_attribute_fn     set_generic_programmable_attribute;
+    sai_get_generic_programmable_attribute_fn     get_generic_programmable_attribute;
+} sai_generic_programmable_api_t;
+```
+
+# API usage #
+
+## Create ##
+
+### APP DB Entry ### 
+```
+127.0.0.1:6379> HGETALL "VIP_TABLE:10.1.1.1"
+1) "nexthop_list"
+2) "10.0.0.57,10.0.0.59"
+3) "ifname"
+4) "Ethernet0, Ethernet1"
+```
+Sample code 
+```c
+sai_attribute_t attr;
+vector<sai_attribute_t> attrs;
+ 
+// set object name
+std::string object_name = “vip_table”;
+attr.id = SAI_GENERIC_PROGRAMMABLE_ATTR_OBJECT_NAME;
+auto size = sizeof(attr.value.chardata);
+strncpy(attr.value.chardata, object_name.c_str(), size - 1);
+attr.value.chardata[size - 1] = '\0';
+attrs.push_back(attr);
+ 
+// encode the JSON info for VIP table
+nlohmann::json json;
+ 
+// push metadata and value for virtual_ip
+nlohmann::json sai_metadata_json;
+sai_metadata_json[“sai_metadata”][“sai_attr_value_type”] = SAI_ATTR_VALUE_TYPE_IP_ADDRESS;
+nlohmann::json virtual_ip;
+virtual_ip[“virtual_ip”][“value”] = “10.1.1.1”;
+virtual_ip[“virtual_ip”][“sai_metadata”] = sai_metadata_json.dump();
+json.push_back(virtual_ip);
+ 
+// push metadata and value for nexthop_list
+nlohmann::json sai_metadata_json;
+sai_metadata_json[“sai_metadata”][“sai_attr_value_type”] = SAI_ATTR_VALUE_TYPE_OBJECT_LIST;
+nlohmann::json nexthop_list;
+nexthop_list [“nexthop_list”][“value”] = { nexthop_1_oid, nexthop_2_iod };
+nexthop_list [“nexthop_list”][“sai_metadata”] = sai_metadata_json.dump();
+json.push_back(nexthop_list);
+ 
+attr.id = SAI_GENERIC_PROGRAMMABLE_ATTR_ENTRY;
+strncpy(attr.data.json, json.dump(), sizeof(json.dump());
+attrs.push_back(attr);
+ 
+sai_object_id_t sai_generic_oid;
+sai_status_t status =
+    sai_generic_programmable_api->create_generic_programmable(
+        &sai_generic_oid, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+```
+
+## Delete ##
+Delete requires no special information since the object is OID based. Just passing the OID should be sufficient to clear the entry.
+
+## Object Query ##
+The list of all generic programmable objects can be queried using the existing saiobject.h infrastructure in SAI.
+
+## CRM usage ##
+```c
+sai_attribute_t attr;
+vector<sai_attribute_t> attrs;
+ 
+// set object name
+std::string object_name = “vip_table”;
+attr.id = SAI_GENERIC_PROGRAMMABLE_ATTR_OBJECT_NAME;
```

Comment:
Attribute based filtering is lost; for eg. let us assume VIP_table has an attribute to say if a given entry is v4 or v6. How do I filter based on v6 entries?